src/mainwin.cc

```diff
@@ -2062,35 +2064,157 @@ void MainWindow::csgRender()
+	if (showDialog) {
+		auto printInitDialog = new PrintInitDialog();
+		auto printInitResult = printInitDialog->exec();
+		printInitDialog->deleteLater();
+		if (printInitResult == QDialog::Rejected) {
+			return;
+		}
+
+		const auto dialog_result = printInitDialog->get_result();
+		print_service = dialog_result.service;
+		if (dialog_result.rememberDecision) {
+			switch (dialog_result.service) {
+			case print_service_t::PRINT_A_THING:
+				s->set(Settings::Settings::printService, "PrintAThing");
+				break;
+			case print_service_t::OCTOPRINT:
+				s->set(Settings::Settings::printService, "OctoPrint");
+				break;
+			default:
+				s->set(Settings::Settings::printService, "None");
+				break;
+			}
+			s->set(Settings::Settings::printServiceShowDialog, false);
+		}
+		Preferences::Preferences::inst()->updateGUI();
+	} else {
+		const auto service = s->get(Settings::Settings::printService).toString();
+		if (service == "PrintAThing") {
+			print_service = print_service_t::PRINT_A_THING;
+		} else if (service == "OctoPrint") {
+			print_service = print_service_t::OCTOPRINT;
+		} else {
+			print_service = print_service_t::NONE;
+		}
+	}
+
+	switch (print_service) {
+	case print_service_t::PRINT_A_THING:
+		PRINT("Sending design to print service Print a Thing");
+		sendToPrintAThing();
+		break;
+	case print_service_t::OCTOPRINT:
+		PRINT("Sending design to OctoPrint");
+		sendToOctoPrint();
+		break;
+	default:
+		if (!showDialog) {
+			PRINT("Sending design to print services is disabled, check Preferences to enable.");
+		}
+		break;
+	}
+#endif
+}
+
+void MainWindow::sendToOctoPrint()
+{
+#ifdef ENABLE_3D_PRINTING
+	Settings::Settings *s = Settings::Settings::inst();
+	const QString fileFormat = QString::fromStdString(s->get(Settings::Settings::octoPrintFileFormat).toString());
+	FileFormat exportFileFormat{FileFormat::STL};
+	if (fileFormat == "OFF") {
+		exportFileFormat = FileFormat::OFF;
+	} else if (fileFormat == "AMF") {
+		exportFileFormat = FileFormat::AMF;
+	} else if (fileFormat == "3MF") {
+		exportFileFormat = FileFormat::_3MF;
+	} else {
+		exportFileFormat = FileFormat::STL;
+	}
+
+	QTemporaryFile exportFile{QDir::temp().filePath("OpenSCAD.XXXXXX." + fileFormat.toLower())};
+	if (!exportFile.open()) {
+		PRINT("Could not open temporary file.");
+		return;
+	}
+	const QString exportFileName = exportFile.fileName();
+	exportFile.close();
+
+	QString userFileName;
+	if (this->fileName.isEmpty()) {
+		userFileName = exportFileName;
+	} else {
+		QFileInfo fileInfo{this->fileName};
+		userFileName = fileInfo.baseName() + "." + fileFormat.toLower();
+	}
+
+	exportFileByName(this->root_geom, exportFileFormat, exportFileName.toLocal8Bit().constData(), exportFileName.toUtf8());
+
+	OctoPrint octoPrint;
+	try {
+		this->progresswidget = new ProgressWidget(this);
+		connect(this->progresswidget, SIGNAL(requestShow()), this, SLOT(showProgress()));
+
+		const QString fileUrl = octoPrint.upload(new QFile(exportFileName), userFileName,
```

Comment:
Is there a reason why you don't construct the QFile instance inside the `upload()` function, as you're handing over ownership to it anyway?
Just nice to make it easy to understand object ownership by keeping the owner closer to the point of creation.

Reply:
I guess I was thinking to change the export stuff so we don't need the temp file but that would have caused too much changes in the end.

changed

Reply:
right, we could ideally provide a QIODevice or smth. if we wanted to stream any info.